src/librustc_typeck/check/demand.rs

```diff
@@ -341,14 +359,22 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                         if let Some(sugg) = self.can_use_as_ref(expr) {
                             return Some(sugg);
                         }
-                        return Some(match mutability {
-                            hir::Mutability::MutMutable => {
-                                (sp, "consider mutably borrowing here", format!("&mut {}",
-                                                                                sugg_expr))
+                        return Some(match (mutability, is_struct_pat_shorthand_field) {
```

Comment:
I'd personally extract this as well... :) `fn field_shorthand_suggestion()` or some such